File: Hashing/1-Two-Sum.js
const twoSum = (arr,target,ans) => {
    for(let i = 0; i < arr.length; i++) {
        let first = arr[i];
        for(j = i + 1; j < arr.length; j++) {
            let second = arr[j];
            if(first + second === target) {
                ans.push(i);
                ans.push(j);
                return ans;
            }
        }
    }
    return [];
}

const twoSumOptimal = (nums, target) => {
    const map = new Map(); // value -> index
    for(let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        if(map.has(complement)) {
            return [map.get(complement),i]; // found the pair
        }
        map.set(nums[i], i); // Store current value and index
    }
    return [];
}

function main() {
    const arr = [2,7,11,15];
    let target = 9;
    let ans = twoSum(arr,target,[]);
    console.log(ans);

    let ansOptimal = twoSumOptimal(arr,target);
    console.log(ansOptimal);
}

main();

File: Hashing/1-Two-Sum.c
#include<stdio.h>
#include<stdlib.h>

#define MAX 1000

void twoSum(int arr[], int n, int target) {
    for(int i = 0; i < n; i++) {
        for(int j = i + 1; j < n; j++) {
            if(arr[i] + arr[j] == target) {
                printf("[%d, %d]", i, j);
                return;
            }
        }
    }
    printf("No Pair Found !");
    return;
}

void twoSumOptimal(int arr[], int n, int target) {
    int hash[MAX] = {0};
    for(int i = 0; i < n; i++) {
        int complement = target - arr[i];
        if(hash[complement] != 0) { // Found the complement
            printf("[%d, %d]\n", hash[complement] - 1, i);
            return;
        }
        hash[arr[i]] = i + 1; // store index+1 to differentiate from 0
    }
    printf("No Pair Found !");
}

int main() {
    int arr[] = {2,7,11,15};
    int target = 18;
    int n = sizeof(arr) / sizeof(arr[0]);

    twoSum(arr, n, target);
    twoSumOptimal(arr, n, target);
}